test/statistics_test.py

```diff
@@ -0,0 +1,161 @@
+# (at your option) any later version.
+#
+# The full text of the license can be found in the file LICENSE in the
+# icepack source directory or at <http://www.gnu.org/licenses/>.
+
+import pytest
+import numpy as np
+import numpy.random as random
+import firedrake
+from firedrake import inner, grad, dx, exp, interpolate, as_vector
+import icepack
+from icepack.statistics import StatisticsProblem, MaximumProbabilityEstimator
+from icepack.constants import (
+    ice_density as ρ_I,
+    water_density as ρ_W,
+    gravity as g,
+    glen_flow_law as n,
+)
+
+
+def test_poisson_problem():
+    Nx, Ny = 32, 32
+
+    mesh = firedrake.UnitSquareMesh(Nx, Ny)
+    degree = 2
+    Q = firedrake.FunctionSpace(mesh, "CG", degree)
+    V = firedrake.FunctionSpace(mesh, "CG", degree)
+
+    x, y = firedrake.SpatialCoordinate(mesh)
+    f = interpolate(firedrake.Constant(1), Q)
+
+    def simulation(q):
+        u = firedrake.Function(V)
+        J = (0.5 * exp(q) * inner(grad(u), grad(u)) - f * u) * dx
+        F = firedrake.derivative(J, u)
+
+        bc = firedrake.DirichletBC(V, 0, "on_boundary")
+        problem = firedrake.NonlinearVariationalProblem(F, u, bc)
+        params = {
+            "solver_parameters": {
+                "pc_type": "lu",
+                "pc_factor_mat_solver_type": "mumps",
+            }
+        }
+        solver = firedrake.NonlinearVariationalSolver(problem, **params)
+        solver.solve()
+
+        return u
+
+    q_true = interpolate(-4 * ((x - 0.5) ** 2 + (y - 0.5) ** 2), Q)
+    u_obs = simulation(q_true)
+
+    def likelihood(u):
+        return 0.5 * (u - u_obs) ** 2 * dx
+
+    α = firedrake.Constant(1e-4)
+
+    def prior(q):
+        return 0.5 * α ** 2 * inner(grad(q), grad(q)) * dx
+
+    q_initial = firedrake.Function(Q)
+    problem = StatisticsProblem(simulation, likelihood, prior, q_initial)
+    estimator = MaximumProbabilityEstimator(problem, gradient_tolerance=1e-7)
+    q = estimator.solve()
+
+    assert firedrake.norm(q - q_true) < 0.25
+
+
+@pytest.mark.parametrize("with_noise", [False, True])
+def test_ice_shelf_inverse(with_noise):
+    Nx, Ny = 32, 32
+    Lx, Ly = 20e3, 20e3
+
+    u0 = 100
+    h0, δh = 500, 100
+    T0, δT = 254.15, 5.0
+    A0 = icepack.rate_factor(T0)
+    δA = icepack.rate_factor(T0 + δT) - A0
+
+    def exact_u(x):
+        ρ = ρ_I * (1 - ρ_I / ρ_W)
+        Z = icepack.rate_factor(T0) * (ρ * g * h0 / 4) ** n
+        q = 1 - (1 - (δh / h0) * (x / Lx)) ** (n + 1)
+        δu = Z * q * Lx * (h0 / δh) / (n + 1)
+        return u0 + δu
+
+    mesh = firedrake.RectangleMesh(Nx, Ny, Lx, Ly)
+    degree = 2
+    V = firedrake.VectorFunctionSpace(mesh, "CG", degree)
+    Q = firedrake.FunctionSpace(mesh, "CG", degree)
+
+    x, y = firedrake.SpatialCoordinate(mesh)
+    u_initial = interpolate(as_vector((exact_u(x), 0)), V)
+    q_initial = interpolate(firedrake.Constant(0), Q)
+    h = interpolate(h0 - δh * x / Lx, Q)
```

Comment:
So this is the old style way of doing things with a continuous thickness field right? Is there good reason not to test this with `h` being a point cloud?

Reply:
This is for consistency with the tests that I had for the old inverse code. I can add a test for point cloud data; do you think that should go in this PR or a later one?

Reply:
Another PR would be better since this is nicely self-contained. 
Note though that you say in the PR description
> This makes it possible to use the Firedrake VertexOnlyMesh features to do sparse data assimilation and to do time-dependent data assimilation
and you've added a tutorial for doing these.
Write issues to get robust tests of these working?